File: 101-200/101-110/105-binaryTreeFromPreInOrder/binaryTreeFromPreInOrder.cpp
#include <vector>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

using namespace std;

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return buildTreeHelper(preorder, 0, inorder, 0, inorder.size());
    }

    TreeNode *buildTreeHelper(vector<int> &preorder, int pre_index, vector<int> &inorder, int left, int right) {
        if (left == right)
            return NULL;

        int middle_value = preorder[pre_index];
        int in_index = find(inorder, middle_value, left, right);
        TreeNode *root = new TreeNode(middle_value);
        root->left = buildTreeHelper(preorder, pre_index+1, inorder, left, in_index);
        root->right = buildTreeHelper(preorder, pre_index+in_index-left+1, inorder, in_index+1, right);
        return root;
    }

    int find(vector<int> &nums, int num, int start, int end) {
        for (int i = start; i < end; ++i)
            if (nums[i] == num)
                return i;

        return -1;
    }
};

File: 101-200/101-110/105-binaryTreeFromPreInOrder/binaryTreeFromPreInOrder.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """

        def buildChildTree(preIndex, inIndex, length):
            if length == 0:
                return None

            root = TreeNode(preorder[preIndex])
            count = 0
            while inorder[inIndex+count] != preorder[preIndex]:
                count += 1

            root.left = buildChildTree(preIndex+1, inIndex, count)
            root.right = buildChildTree(preIndex+count+1, inIndex+count+1, length-count-1)
            return root

        return buildChildTree(0, 0, len(preorder))

    def buildTree2(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """

        if not preorder:
            return None

        root = TreeNode(preorder[0])
        index = inorder.index(preorder[0])
        root.left = self.buildTree(preorder[1:index+1], inorder[:index])
        root.right = self.buildTree(preorder[index+1:], inorder[index+1:])
        return root